drivers/video/video_ctrls.c

```diff
@@ -0,0 +1,241 @@
+{
+	switch (type) {
+	case VIDEO_CTRL_TYPE_BOOLEAN:
+		if (step != 1 || max > 1 || min < 0) {
+			return -ERANGE;
+		}
+		return 0;
+	case VIDEO_CTRL_TYPE_INTEGER:
+	case VIDEO_CTRL_TYPE_INTEGER64:
+		if (step == 0 || min > max || def < min || def > max) {
+			return -ERANGE;
+		}
+		return 0;
+	case VIDEO_CTRL_TYPE_MENU:
+	case VIDEO_CTRL_TYPE_INTEGER_MENU:
+		if (min > max || def < min || def > max || min < 0) {
+			return -ERANGE;
+		}
+		return 0;
+	default:
+		return 0;
+	}
+}
+
+static inline void set_type_flag(uint32_t id, enum video_ctrl_type *type, uint32_t *flags)
+{
+	*flags = 0;
+
+	switch (id) {
+	case VIDEO_CID_HFLIP:
+	case VIDEO_CID_VFLIP:
+		*type = VIDEO_CTRL_TYPE_BOOLEAN;
+		break;
+	case VIDEO_CID_POWER_LINE_FREQUENCY:
+	case VIDEO_CID_TEST_PATTERN:
+		*type = VIDEO_CTRL_TYPE_MENU;
+		break;
+	case VIDEO_CID_PIXEL_RATE:
+		*type = VIDEO_CTRL_TYPE_INTEGER64;
+		*flags |= VIDEO_CTRL_FLAG_READ_ONLY;
+		break;
+	default:
+		*type = VIDEO_CTRL_TYPE_INTEGER;
+		break;
+	}
+}
+
+int video_ctrl_init(struct video_ctrl *ctrl, struct video_ctrl_handler *hdl, uint32_t id,
+		    int32_t min, int32_t max, uint32_t step, int32_t def)
+{
+	int ret;
+	enum video_ctrl_type type;
+	uint32_t flags;
+
+	set_type_flag(id, &type, &flags);
+
+	/* Sanity checks */
+	if (id == 0 || id >= VIDEO_CID_PRIVATE_BASE) {
+		return -EINVAL;
+	}
+
+	ret = check_range(type, min, max, step, def);
+	if (ret) {
+		return ret;
+	}
+
+	ctrl->handler = hdl;
+
+	ctrl->id = id;
+	ctrl->type = type;
+	ctrl->flags = flags;
+	ctrl->min = min;
+	ctrl->max = max;
+	ctrl->step = step;
+	ctrl->def = def;
+	ctrl->val = def;
+
+	sys_slist_append(&hdl->list, &ctrl->node);
+
+	return 0;
+}
+
+struct video_ctrl *video_ctrl_find(struct video_ctrl_handler *hdl, uint32_t id)
+{
+	struct video_ctrl *ctrl;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&hdl->list, ctrl, node) {
+		if (ctrl->id == id) {
+			return ctrl;
+		}
+	}
+
+	return NULL;
+}
+
+static int video_ctrl_find_by_root_dev(const struct device *dev, uint32_t id,
+				       struct video_ctrl **ctrl)
+{
+	struct video_device *vdev = video_find_vdev(dev);
+
+	if (!vdev) {
+		return -EINVAL;
+	}
+
+	struct video_ctrl_handler *hdl = &vdev->ctrl_handler;
+
+	if (!hdl) {
+		LOG_ERR("Control handler not found\n");
+		return -ENOTSUP;
+	}
+
+	*ctrl = video_ctrl_find(hdl, id);
+	if (!*ctrl) {
+		LOG_ERR("Control id 0x%x not found\n", id);
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+int video_get_ctrl(const struct device *dev, struct video_control *control)
+{
+	struct video_ctrl *ctrl;
+
+	int ret = video_ctrl_find_by_root_dev(dev, control->id, &ctrl);
+
+	if (ret) {
+		return ret;
+	}
+
+	if (ctrl->flags & VIDEO_CTRL_FLAG_WRITE_ONLY) {
+		LOG_ERR("Control id 0x%x is write-only\n", control->id);
+		return -EACCES;
+	}
+
+	control->val = ctrl->val;
+
+	return 0;
+}
+
+int video_set_ctrl(const struct device *dev, struct video_control *control)
+{
+	struct video_ctrl *ctrl;
+
+	int ret = video_ctrl_find_by_root_dev(dev, control->id, &ctrl);
```

Comment:
I wonder how to implement video aggregation with this:
```
[imager0]---MIPI--.
                 [aggregator0]---MIPI---[mipirx0]---RAM---> application
[imager1]---MIPI--'
```

If the `aggregator0` produces a single feed with `imager0` and `imager1` output side-by-side with some overlay, there could be a different exposure and gain to compensate i.e. strong light coming from one side.

How would the application tell "tune the left image exposure"?

Reply:
`aggregator0` could be an external chip too, sending a MIPI feed to `mipirx0`. FPGA are often used for video aggregation.

For instance, connecting 10 sensors in a line, with a single i.MXRT1170 mipi DMA engine.

Reply:
I think we should base on realistic usecases. Could you give a real example (HW and SW) that have multiple sensors connects to a single MIPI CSI2Rx receiver (this is common as a receiver can support multiple sensor) and **a single DMA engine** (this is  uncommon as I never see a HW like that). Note that the DMA engine is not the same as the MIPI CSI2 receiver. If we have a single DMA engine, it has to do multiplexing to dump images from each sensor to RAM ? I hadn't see a HW like that and wonder it is supported, for example in Linux ?

Reply:
Lattice uses the CrossLink-NX series for this a lot:
- https://www.latticesemi.com/en/Products/DesignSoftwareAndIP/IntellectualProperty/IPCore/IPCores04/4Inputto1OutputMIPICSI2CameraAggregatorBridge
- https://www.latticesemi.com/products/designsoftwareandip/intellectualproperty/referencedesigns/referencedesign04/4to1
- https://www.latticesemi.com/en/Products/DesignSoftwareAndIP/IntellectualProperty/ReferenceDesigns/ReferenceDesign04/Nto1

Example of devkit demoing this purpose:

- https://www.latticesemi.com/products/developmentboardsandkits/crosslink-nx_vip_sensor_input

Here is one from AntMicro which runs Zephyr even:

- https://antmicro.com/blog/2022/12/sdi-mipi-video-converter/

It seems like Sony tries to avoid needing this part by doing astonishingly *wide* image sensors ^_^

- https://www.sony-semicon.com/en/products/is/industry/gs/imx901-902.html

Some other examples of vendors using this:

- https://www.e-consystems.com/ip-cores/vision-aggregator-ipcore.asp
- https://www.nuvation.com/multi-camera-input-aggregation-device
- https://www.macnica.co.jp/en/business/semiconductor/articles/lattice/135699/
- https://www.edge-ai-vision.com/2023/06/gmsl-cameras-vs-mipi-cameras-a-detailed-study/
- https://www.vision-components.com/en/news/embedded-world-ultracompact-stereo-camera-and-new-mipi-modules/
- https://www.arducam.com/camarray-release-multiple-mipi-camera-solution/
- https://www.e-consystems.com/nvidia-cameras/jetson-agx-xavier-cameras/stereo-camera.asp
- https://camerafmc.com/

Reply:
> dev routing mechanism and API
> How controls can be distinguished for each sensor ?

The basic scheme that was there before required some boilerplate for forwarding the controls to the next device(s):

```c
static int video_set_ctrl(const struct device *dev, unsigned int cid, void *value)
{
        const struct videodriver_config *cfg = dev->config;

        return video_set_ctrl(dev->source_dev);
}
```

Something custom could easily be made:

```c
static int aggregator_set_ctrl(const struct device *dev, unsigned int cid, void *value)
{
        const struct aggregator_config *cfg = dev->config;
        float adjust = 1;
        int value_left, value_right;
        int ret;

        switch (cid) {
        case VIDEO_CID_EXPOSURE:
                aggregator_get_exposure_correction(dev, &adjust);
                break;
         case VIDEO_CID_GAIN:
                aggregator_get_gain_correction(dev, &adjust);
                break;

         }

        value_left = (int)value * adjust;
        value_right = (int)value * (1.0 - adjust);

        ret = video_set_ctrl(dev->source_dev_left, cid, (void *)value_left);
        if (ret < 0) {
                return ret;
        }

        ret = video_set_ctrl(dev->source_dev_right, cid, (void *)value_right);
        if (ret < 0) {
                return ret;
        }

        return 0;
}
```

Maybe just a little bit of tuning permits to preserve these kinds of setup.

I am curious about how this is handled Linux-side too! I will look it up.

Reply:
And we can already see the benefit the current PR introducing types associated with controls. Here I had the wrong assumption that all controls were `(int)`, which can only be done thanks to `ctrl->type` :)

Reply:
> Have a look at this similar usecase about virtual channels / multiple-stream
> https://linuxtv.org/downloads/presentations/media_summit_2016_san_diego/v4l2-multistream.pdf

These "multiplexer" might very well be one of the solutions presented above. Thank you, very on-topic! I will review it.

I will take some more time to study it along with the PR, this might become more clear to me then.
Thank you once again to take the time to explain this!

Reply:
> These "multiplexer" might very well be one of the solutions presented above.

Sorry, maybe I was unclear. In fact, Linux already supports multi-sensors scenarios and to do this, as being said, we have to aggregate controls from different sensors into different video devices (`/dev/videoX` in Linux). I quoted the above slides just to show this idea easier where it is written in text (as it is different to show it in complex Linux code) but it does not mean this v4l2 extension proposal is accepted in Linux (it was not).

I will come back to this PR when possible to answer your remaining questions and with some added documentation, hope it will be clearer.

Reply:
> different sensors into different video devices

If it is still possible to call the APIs on an individual driver, that might solve a lot of low-level problems... Allowing to add software that enable to implement the pipeline.

> maybe I was unclear

That paper and your explanations are clear, it just takes me time to grasp the whole model... ^_^' which was built over many busy years by you and all V4L2 team!

> don't need to control each sensor separately as the two sensors need to be synced though

Maybe a 360 camera with 2 sensors would be a good example: With a strong sun, that would make a lot of contrast: some side would have direct sunlight and all the shady face of the scene, and the other reflected light from the sun.

![Untitled](https://github.com/user-attachments/assets/8dcb9ec3-7ced-4877-9b47-ad06fe8d1d75)

Reply:
Some more edge cases, which seems to require to allow talking directly to a particular driver...

**Merging streams in one with hardware**

Like lens shading being a little bit different for every camera module of a many-camera system... and a firmware detecting it and telling the ISP to compensate more or less. Also, tuning the Region Of Interest (ROI) differently for every sensor, in order to align the multiple feeds so that they overlap nicely.

**Splitting a single video source**

[Light Field Cameras](https://cs.brown.edu/courses/csci1290/labs/lab_lightfields/images/cameraarrays.png) can be implemented as many individual cameras (left) or through optics (right). and then properties of a single sensor would affect many individual output streams, if they are to be split with hardware (i.e. for debugging/inspection).

**This is a non-problem though**

As it is always possible to call `((struct video_api*)dev->api)->set_ctrl()` directly for the most exotic use-cases.